src/lib/focus-handler.js

```diff
@@ -0,0 +1,206 @@
+import * as React from "react"
+import { useFocusContext } from "./hooks-context"
+import { FocusContext } from "./hooks-create-context"
+
+export const FocusHandler = ({ uri, location, component, ...domProps }) => {
+  const requestFocus = useFocusContext()
+
+  return (
+    <FocusHandlerImpl
+      {...domProps}
+      component={component}
+      requestFocus={requestFocus}
+      uri={uri}
+      location={location}
+    />
+  )
+}
+
+const FocusHandlerImpl = ({
+  children,
+  style,
+  requestFocus,
+  component: Comp = "div",
+  uri,
+  location,
+  ...domProps
+}) => {
+  let node = React.useRef()
+
+  const [shouldFocus, setShouldFocus] = React.useState(true)
+  const [prevUri, setPrevUri] = React.useState(uri)
+  const [prevPathname, setPrevPathname] = React.useState(location.pathname)
+  const [focusHandlerCount, setFocusHandlerCount] = React.useState(0)
+  const initialRender = React.useRef(true)
+
+  if (uri) {
+    const myURIChanged = uri !== prevUri
+    const navigatedUpToMe =
+      prevPathname !== location.pathname && location.pathname === uri
+
+    setPrevUri(uri)
+    setPrevPathname(location.pathname)
+
+    setShouldFocus(myURIChanged || navigatedUpToMe)
+  }
+
+  React.useEffect(() => {
+    setFocusHandlerCount(f => f + 1)
+    focus()
+
+    return () => {
+      setFocusHandlerCount(f => f - 1)
+      if (focusHandlerCount === 0) {
+        initialRender.current = true
+      }
+    }
+  }, [])
```

Comment:
You need to keep focusHandlerCount outside of React as this is being tracked globally
```suggestion
  React.useEffect(() => {
    focusHandlerCount++;
    focus()

    return () => {
      focusHandlerCount--;
      if (focusHandlerCount ===  0) {
        initialRender.current = true
      }
    }
  }, [])
```